## Quadratic Formula
MATH 420 <br>
_Spring 2021_ <br>

Let's write a Julia function that finds the roots of a quadratic equation; easy-peasy right? Wrong.  

We start with the formula that we all know and love: 

In [1]:
function solve_quadratic1(a,b,c)
    d = sqrt(b^2 - 4*a*c)
    (-b + d)/(2*a), (-b - d)/(2*a)
end   

solve_quadratic1 (generic function with 1 method)

Possibly we should special case $a = 0$ so that we don't divide by zero, but we'll leave that for another day.

Let's _manufacture_ coefficients with a known roots; say the solutions are $r_1$  and $1/r_1$, where

In [13]:
r1 = 8.0e7;

In [14]:
r2 = 1/r1;

Expanding $(x - r_1) (x - r_2) = x^2 - (r_1+r_2) x + r_1 r_2$, we see the coefficients are

In [15]:
(a,b,c) = (1,-r1-r2,r1*r2)

(1, -8.000000000000001e7, 1.0)

The solutions are

In [16]:
solve_quadratic1(a,b,c)

(8.0e7, 1.4901161193847656e-8)

One solution is $1.25 \times 10^{-8}$, but we get $1.4901161193847656 \times 10^{-8}$. Oh, that's _terrible_. Using the alternative quadratic formula with the radicals in the denominator, our function is

In [9]:
function solve_quadratic2(a,b,c)
    d = sqrt(b^2 - 4*a*c)
    2*c/(-b - d), 2*c/(-b + d)
end   

solve_quadratic2 (generic function with 1 method)

This time we get an accurate value for $1.25 \times 10^{-8}$, but not for the root $8.0 \times 10^{7}$. This is very perverse-both methods give _one_ inaccurate result

In [10]:
solve_quadratic2(a,b,c)

(6.7108864e7, 1.25e-8)

The problem is with an ill-conditioned sum $-b + d$ or $-b - d$. Since $d$ is positive (it's the output of a square root), the sum
$-b + d$ is possibly ill-conditioned when $b > 0$. And when $b < 0$, the sum $-b - d$ is possibly ill-conditioned. 

To simplify the logic, let's negate all the coefficients when $b < 0$. After that we know that $b$ is positive, so the sum $-b - d$ is well-conditioned. This gives

In [11]:
function solve_quadratic3(a,b,c)
    if b < 0 (a,b,c) = (-a,-b,-c) end #now b > 0
    d = sqrt(b^2 - 4*a*c)
    (-b - d)/(2*a), 2*c/(-b-d) # all sums well conditioned
end   

solve_quadratic3 (generic function with 1 method)

Now both roots are accurate!

In [12]:
 solve_quadratic3(a,b,c)

(8.0e7, 1.25e-8)

This is a very famous example of an ill-conditioned sum (aka subtractive cancellation); for references, see, for example

1. https://en.wikipedia.org wiki/Quadratic_equation#Avoiding_loss_of_significance. 

2. http://i.stanford.edu/pub/cstr/reports/cs/tr/66/40/CS-TR-66-40.pdf


There are additional problems when $b^2 - 4ac$ is ill-conditioned.

What about solving a cubic?  The solutions to $x^3 + b x^2 + c x + d = 0$ are

$$\left[ x=\left({{-1}\over{2}}-{{\sqrt{3}\,i}\over{2}}\right)\,
 \left({{\sqrt{27\,d^2+\left(4\,b^3-18\,b\,c\right)\,d+4\,c^3-b^2\,c^
 2}}\over{2\,3^{{{3}\over{2}}}}}+{{b\,c-3\,d}\over{6}}+{{\left(-1
 \right)\,b^3}\over{27}}\right)^{{{1}\over{3}}}-{{\left({{\sqrt{3}\,i
 }\over{2}}+{{-1}\over{2}}\right)\,\left({{c}\over{3}}+{{\left(-1
 \right)\,b^2}\over{9}}\right)}\over{\left({{\sqrt{27\,d^2+\left(4\,b
 ^3-18\,b\,c\right)\,d+4\,c^3-b^2\,c^2}}\over{2\,3^{{{3}\over{2}}}}}+
 {{b\,c-3\,d}\over{6}}+{{\left(-1\right)\,b^3}\over{27}}\right)^{{{1
 }\over{3}}}}}+{{\left(-1\right)\,b}\over{3}} , x=\left({{\sqrt{3}\,i
 }\over{2}}+{{-1}\over{2}}\right)\,\left({{\sqrt{27\,d^2+\left(4\,b^3
 -18\,b\,c\right)\,d+4\,c^3-b^2\,c^2}}\over{2\,3^{{{3}\over{2}}}}}+{{
 b\,c-3\,d}\over{6}}+{{\left(-1\right)\,b^3}\over{27}}\right)^{{{1
 }\over{3}}}-{{\left({{-1}\over{2}}-{{\sqrt{3}\,i}\over{2}}\right)\,
 \left({{c}\over{3}}+{{\left(-1\right)\,b^2}\over{9}}\right)}\over{
 \left({{\sqrt{27\,d^2+\left(4\,b^3-18\,b\,c\right)\,d+4\,c^3-b^2\,c^
 2}}\over{2\,3^{{{3}\over{2}}}}}+{{b\,c-3\,d}\over{6}}+{{\left(-1
 \right)\,b^3}\over{27}}\right)^{{{1}\over{3}}}}}+{{\left(-1\right)\,
 b}\over{3}} , x=\left({{\sqrt{27\,d^2+\left(4\,b^3-18\,b\,c\right)\,
 d+4\,c^3-b^2\,c^2}}\over{2\,3^{{{3}\over{2}}}}}+{{b\,c-3\,d}\over{6
 }}+{{\left(-1\right)\,b^3}\over{27}}\right)^{{{1}\over{3}}}-{{{{c
 }\over{3}}+{{\left(-1\right)\,b^2}\over{9}}}\over{\left({{\sqrt{27\,
 d^2+\left(4\,b^3-18\,b\,c\right)\,d+4\,c^3-b^2\,c^2}}\over{2\,3^{{{3
 }\over{2}}}}}+{{b\,c-3\,d}\over{6}}+{{\left(-1\right)\,b^3}\over{27
 }}\right)^{{{1}\over{3}}}}}+{{\left(-1\right)\,b}\over{3}} \right] 
 $$
 
I doubt that anybody has ever figured out how to alter these solutions to make them safe for numerical evalution. If you need to solve a cubic equation numerically, the exact representations for the solutions is a _terrible_ option.
